zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/S3NotebookRepo.java

```diff
@@ -70,69 +74,95 @@
+    }
+    else {
+      // regular S3
+      this.s3client = new AmazonS3Client(credentialsProvider);
+    }
+
+    // set S3 endpoint to use
+    s3client.setEndpoint(conf.getEndpoint());
+  }
+
+  /**
+   * Create an instance of a custom encryption materials provider class
+   * which supplies encryption keys to use when reading/writing data in S3.
+   */
+  private EncryptionMaterialsProvider createCustomProvider(ZeppelinConfiguration conf) {
+    // use a custom encryption materials provider class
+    String empClassname = conf.getS3EncryptionMaterialsProviderClass();
+    EncryptionMaterialsProvider emp;
+    try {
+      Object empInstance = Class.forName(empClassname).newInstance();
+      if (empInstance instanceof EncryptionMaterialsProvider) {
+        emp = (EncryptionMaterialsProvider) empInstance;
+      }
+      else {
+        throw new IllegalArgumentException("Class " + empClassname + " does not implement "
+                + EncryptionMaterialsProvider.class.getName());
+      }
+    }
+    catch (Exception e) {
+      throw new RuntimeException("Unable to instantiate encryption materials provider class "
+              + empClassname + ": " + e, e);
+    }
+
+    return emp;
   }
 
   @Override
   public List<NoteInfo> list() throws IOException {
-    List<NoteInfo> infos = new LinkedList<NoteInfo>();
-    NoteInfo info = null;
+    List<NoteInfo> infos = new LinkedList<>();
+    NoteInfo info;
     try {
       ListObjectsRequest listObjectsRequest = new ListObjectsRequest()
-          .withBucketName(bucketName)
-          .withPrefix(user + "/" + "notebook");
+              .withBucketName(bucketName)
+              .withPrefix(user + "/" + "notebook");
       ObjectListing objectListing;
       do {
         objectListing = s3client.listObjects(listObjectsRequest);
 
         for (S3ObjectSummary objectSummary :
-          objectListing.getObjectSummaries()) {
+                objectListing.getObjectSummaries()) {
           if (objectSummary.getKey().contains("note.json")) {
             try {
               info = getNoteInfo(objectSummary.getKey());
               if (info != null) {
                 infos.add(info);
               }
-            } catch (AmazonServiceException ase) {
-              LOG.warn("Caught an AmazonServiceException for some reason.\n" +
-                  "Error Message: {}", ase.getMessage());
-            } catch (AmazonClientException ace) {
-              LOG.info("Caught an AmazonClientException, " +
-                  "which means the client encountered " +
-                  "an internal error while trying to communicate" +
-                  " with S3, " +
-                  "such as not being able to access the network.");
-              LOG.info("Error Message: " + ace.getMessage());
-            } catch (Exception e) {
-              LOG.error("Can't read note ", e);
+            } catch (IOException e) {
```

Comment:
My reasoning here is that any of the AWS exceptions actually have quite a bit of detail about what went wrong, so the wording was unnecessary.  AmazonServiceException extends AmazonClientException, so the new code still catches everything and logs an error.   Given this all is ultimately called from NotebookServer, which doesn't log a stack trace on IOException I felt it better to log an error here.

Honestly, I think logging an error here and then throwing an IOException would be better